torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -469,7 +472,20 @@ void RequestCallbackImpl::processRpc(
           [responseFuture,
            messageId,
            fromWorkerId,
-           weak = std::weak_ptr<FutureMessage>(wrappedRpcResponseFuture)]() {
+           weak = std::weak_ptr<FutureMessage>(wrappedRpcResponseFuture),
+           ctxId = autogradContext->contextId()]() {
+            // As this callback can be invoked by a different thread, we have to
+            // make sure that the thread_local states in the previous thread is
+            // correctly propagated.
+            // NB: The execution of TorchScript functions can also run on a
+            // different thread, which is addressed by
+            // https://github.com/pytorch/pytorch/pull/36395
+            // NB: when adding async UDF support, we should also propagate
+            // thread_local states there.
+            // TODO: Land on a general solution for RPC ThreadLocalState. See
+            // https://github.com/pytorch/pytorch/issues/38510
+            DistAutogradContextGuard ctxGuard(ctxId);
```

Comment:
Shouldn't we do this in `addCallback` itself instead of fixing each callsite of addCallback?

Reply:
Yes, eventually, we should do that I think. But for now we still have two Futures (utils and ivalue), and not every callback needs to capture ThreadLocalState, it might be better to first fix master for now. Let's revisit this when we reach a consensus on how we should implement RPC ThreadLocalState.